keps/sig-network/20190603-EndpointSlice-API.md

```diff
@@ -0,0 +1,273 @@
+
+### Kube-Proxy
+
+Watch: Service, EndpointSlice ==> Manage: iptables, ipvs, etc
+
+- Merge multiple EndpointSlice into an aggregated list.
+- Reuse the existing processing logic 
+
+### Endpoint Controller (classic)
+In order to ensure backward compatibility for external consumer of the core/v1 Endpoints API, the existing K8s endpoint controller will keep running until the API is EOL. The following limitations will apply:
+
+- Starting from EndpointSlice beta: If # of endpoints in one Endpoints object exceed 100, generate a warning event to the object. 
+- Starting from EndpointSlice GA: Only include up to 500 endpoints in one Endpoints Object. 
+
+## Roll Out Plan
+
+| K8s Version | State | OSS Controllers                                                         | Internal Consumer (Kube-proxy) |
+|-------------|-------|-------------------------------------------------------------------------|--------------------------------|
+| 1.16        | Alpha | EndpointSliceController (Alpha) EndpointController (GA)                 | Endpoints                      |
+| 1.17        | Beta  | EndpointSliceController (Beta)  EndpointController (GA with warning)    | EndpointSlice                  |
+| 1.18        | GA    | EndpointSliceController (GA)    EndpointController (GA with limitation) | EndpointSlice                  |
+
+
+
+## FAQ
+
+- Why only include up to 100 endpoints in one EndpointSlice object? Why not 1 endpoint? Why not 1000 endpoints?
+
+Based on the data collected from user clusters, vast majority (> 99%) of the k8s services have less than 100 endpoints. For small services, EndpointSlice API will make no difference. If the MaxEndpointThreshold is too small (e.g. 1 endpoint per EndpointSlice), controller loses capability to batch updates, hence causing worse write amplification on service creation/deletion and scale up/down. Etcd write RPS is significant limiting factor.
+
+- Why do we have a status struct for each endpoint? Why not boolean state for readiness?
+
+The current Endpoints API only includes a boolean state (Ready vs. NotReady) on individual endpoint. However, according to pod life cycle, there are more states (e.g. Graceful Termination, ContainerReary). In order to represent additional states other than Ready/NotReady,  a status structure is included for each endpoint. More condition types can be added in the future without compatibility disruptions. As more conditions are added, different consumer (e.g. different kube-proxy implementations) will have the option to evaluate the additional conditions. 
+
+
+## Graduation Criteria
+
+In order to graduate to beta, we need:
+
+- Kube-proxy switch to consume EndpointSlice API. 
+- Verify performance/scalability via testing.
+
+## Alternatives
+
+- increase the etcd size limits
+- endpoints controller batches / rate limits changes
+- apiserver batches / rate-limits watch notifications
+- apimachinery to support object level pagination
```

Comment:
I think more discussion of the alternatives, and why you think EndpointSlice is the best one, would be useful. Are any other parts of kubernetes running into scaling issues that might be fixable with one of these more-generic fixes?

Another alternative would be to have kube-proxy just ignore Endpoints, and instead watch Pods and compute the endpoints itself. (This has similar properties to "1 Endpoint per EndpointSlice" in some ways but not in others.)

Reply:
^ Agreed, this is a brand new pattern, and would set a substantial precedent.

Reply:
Any of the generic solutions would really solve the problem - they just help to some extent, I've heard users asking for.

Watching all pods by all nodes would kill us :)

Reply:
> Watching all pods by all nodes would kill us :)

well, note that some (many?) network plugins have to do that to implement NetworkPolicy currently, so a more generic fix for the Endpoints problem would help with NetworkPolicy scaling too.

Reply:
We have briefly explored the alternatives. The batch/rate limiting mechanisms essentially trade supporting large # of endpoints with something like endpoint update latency. In order to achieve optimal result, it can get quite complex when more dimensions are considered, like # of nodes etc. Hence, it is equal or more work for apimachinery to implement these features, with less performance gain. 

In addition, Slicing up endpoints aligns better with the established pattern for load balancer scalability: endpoint subsetting. Hence, open the door for future extension in this direction.

Reply:
I feel like scalability for network policy  is very different from endpoints scalability. It heavily depends on the underlying implementation. 

The worst case is "Watching all pods by all nodes". 
The ideal case is like "Only watch pods on its own node". 

So I do not think this particular proposal would help in that sense. 

However, if one day we decided to go with service policy, then this proposal will automatically benefit the implementation.

Reply:
So we almost discussed this in the SIG meeting last night but were then running out of time. Anyway, I found `EndpointSlice` to be a fairly surprising proposal, and not how I would have guessed that people would have wanted to address this issue, so I wanted to see more discussion (in the KEP, documented for posterity) about why you ended up going with this approach rather than the alternatives.

Reply:
added `why not pursue the alternatives?` in the FAQ section